1218 check delegate wallets

devyetii opened the pull request:
## Fixes
- #1218
- Delegate wallets and operational wallet of a provider should differ
## Changes
- Add checks for this condition in the registration phase of the providers in the corresponding smart contracts.

## Need to be mentioned in CHANGELOG.md?
- Yes

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

devyetii committed:
Comparing delegate wallet with client_id corresponding to miner's public key

devyetii committed:
Disable delegate wallet checking  for develoment

devyetii committed:
Add delegate wallet check to sharder

devyetii committed:
Add delegate wallet check to blobber

devyetii committed:
Merge branch 'staging' into 1218-check-delegate-wallets

devyetii committed:
Add delegate wallet check for validator

devyetii committed:
Add delegate wallet check to authorizer

devyetii committed:
Fix logger problem in authorizer

devyetii commented:
@dabasov Kindly check

peterlimg commented:
@devyetii is this still WIP, or is ready to review

devyetii commented:
@peterlimg waiting for review from @dabasov

peterlimg commented on code/go/0chain.net/smartcontract/minersc/miner.go:
```diff
@@ -84,6 +87,27 @@ func (msc *MinerSmartContract) AddMiner(t *transaction.Transaction,
+		if err != nil {
+			logging.Logger.Error("Couldn't decode public key to compare to delegate wallet")
+			return "", common.NewError("add_miner",
+				"Couldn't decode publick key to compare to delegate wallet")
```
Let's avoid capital error message

devyetii replied:
Done

peterlimg commented on code/go/0chain.net/smartcontract/minersc/miner.go:
```diff
@@ -84,6 +87,27 @@ func (msc *MinerSmartContract) AddMiner(t *transaction.Transaction,
+		if operationalClientID == newMiner.Settings.DelegateWallet {
+			logging.Logger.Error("Can't use the same wallet as both operational and delegate")
+			return "", common.NewError("add_miner",
+				"Can't use the same wallet as both operational and delegate")
```
Avoid capital error message, and use consistent message, so change "Can't" to "couldn't" or vice verse. Also I saw repeat code below that basically do the same thing, which we can improve by moving this code to a function, and define in parent package `code/go/0chain.net/smartcontract`.

devyetii replied:
Done

devyetii committed:
Merge branch 'staging' into 1218-check-delegate-wallets

devyetii committed:
Create a single function to validate wallet and use it for all the providers

peterlimg commented on code/go/0chain.net/smartcontract/common/wallet.go:
```diff
@@ -0,0 +1,27 @@
+			"couldn't use the same wallet as both operational and delegate")
+	}
+
+	return nil
```
Oh, actually this functions will be reused somewhere else and should be put in the `encryption` package, where we do the public/private/client ID things. I've done that in another [PR](https://github.com/0chain/0chain/pull/1850/files#diff-a4b5842743d9fb9c0c1e8f9d21e4b238ccf11fa0cb59fef25a8890ce57180622R176-R188) but not merged it, let me check if we can merge it today.

devyetii commented:
@peterlimg  Sorry for the late reply. I've added a reply to your comment in the same day but actually wasn't added and it's still "Pending" and I don't know why 

Kindly note that this function (the one I added in this PR) doesn't check if the public key and the wallet ID matches, otherwise it checks that the operational wallet ID of the provider (the ID generated by transforming the PK of the provider, be it a miner, sharder, blobber, validator or authorizer) and the delegate wallet given as input to the registration function of this provider doesn't match. If they match, an error should be returned.

devyetii committed:
Merge branch 'staging' into 1218-check-delegate-wallets

devyetii committed:
Merge branch 'staging' into 1218-check-delegate-wallets

dabasov commented on code/go/0chain.net/smartcontract/common/wallet.go:
```diff
@@ -0,0 +1,27 @@
+
+
+func ValidateWallet(publicKey, delegateWalletID string) *common.Error {
+	publicKeyBytes, err := hex.DecodeString(publicKey)
```
why do you decode? there should be method to get walletId from pub

dabasov commented on code/go/0chain.net/smartcontract/storagesc/blobber.go:
```diff
@@ -200,7 +203,7 @@ func (sc *StorageSmartContract) removeBlobber(t *transaction.Transaction,
 // tokens left will be moved to unlocked part of related stake pool;
 // the part can be moved back to the blobber anytime or used to
-// increase blobber's capacity or write_price next time
+// increase blobber's- capacity or write_price next time
```
remove

dabasov commented on code/go/0chain.net/smartcontract/minersc/sharder.go:
```diff
@@ -113,6 +115,13 @@ func (msc *MinerSmartContract) AddSharder(
 	}
 
+	// Check delegate wallet differs from operationl wallet
+	if ! config.Development() {
```
put this check into ValidateWallet for now

devyetii committed:
Remove unexpected character

devyetii committed:
Move development mode check to ValidateWallet + Function to get walletID from PK

devyetii committed:
Merge branch 'staging' into 1218-check-delegate-wallets

dabasov reviewed (approved):
lgtm

devyetii committed:
Try to fix circular dependency in tests

devyetii committed:
Try to fix circular dependency in tests 2

peterlimg commented on code/go/0chain.net/smartcontract/common/wallet.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+/*GetWalletIdFromPublicKey - given the PK of the provider, return the its operational walletID i.e. the key used to sign the txns */
+func GetWalletIdFromPublicKey(pk string) (string, error) {
```
Let's move this function into `encryption` package, and name it as `GetClientIDFromPublickKey()`, comment with `//` instead of `/**/`.

peterlimg commented on code/go/0chain.net/smartcontract/common/wallet.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+
+func ValidateWallet(publicKey, delegateWalletID string) *common.Error {
```
Rename to `ValidateDelegateWallet()` and add comments to describe what it does, we need to ensure it would not be misused by users, i.e thought the the function will return nil if the clientID and pubkey matches

devyetii committed:
Move client id from pk converter into encruption package

devyetii committed:
Merge branch 'staging' into 1218-check-delegate-wallets

devyetii committed:
Use the renamed function

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 1218-check-delegate-wallets  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg commented on code/go/0chain.net/core/encryption/helpers.go:
```diff
@@ -0,0 +1,13 @@
+import "encoding/hex"
+
+// GetClientIDFromPublickKey - given the PK of the provider, return the its operational walletID i.e. the key used to sign the txns
+func GetClientIDFromPublickKey(pk string) (string, error) {
```
`*PublickKey` -> `*PublicKey`

peterlimg commented on code/go/0chain.net/core/encryption/helpers.go:
```diff
@@ -0,0 +1,13 @@
+	}
+	operationalClientID := Hash(publicKeyBytes)
+	return operationalClientID, nil
+}
```
We can put the function in keys.go, it's not a helper function actually.

devyetii committed:
Fix typo + move transformer fn

devyetii committed:
Merge branch 'staging' into 1218-check-delegate-wallets

dabasov reviewed (approved):
lgtm
